README.md
## Bitmap font generator for msdf-gdx
This is a small utility used to create SDF, MSDF and MSDFA bitmap font files from `.ttf` files.
The generated `.fnt` files uses libGDX's format. 

Here's what you need to download:
- The latest version of Chlumsky/msdfgen can be found **[here][msdfgen]**.
- The utility jar file can be found in the **[releases][releases]**.

Place them both in the same directory.

### Usage
The jar file can be run with:
```text
java -jar msdfgen.jar [options] <font-file>
```
You can also use a configuration file `config.txt` with:
```text
java -jar msdfgen.jar @config.txt 
```
The CLI expects at least one parameter specifying the `.ttf` font file, 
but there can be more than one.

#### Options
- **`-g`**, **`--msdfgen`**: Path of the msdfgen executable. Default: `msdfgen.exe`.
- **`-o`**, **`--output`**: Output path of generated font textures. Default: same directory.
- **`-t`**, **`--field-type`**: Field type, one of: sdf, psdf, msdf, mtsdf. Default: msdf.
If mtsdf is used, alpha field type option is ignored.
- **`-a`**, **`--alpha-field-type`**: Alpha field type, one of: none, sdf, psdf. Default: sdf.
- **`-s`**, **`--font-size`**: Font size for generated textures. Default: 32.
- **`-r`**, **`--distance-range`**: Distance range in which SDF is encoded. Default: 5.
- **`-d`**, **`--texture-size`**: Maximum width and height of generated atlas pages. Default: 512, 512.
- **`-p`**, **`--padding`**: Padding between glyphs and on the border of the atlas pages. Default: 2.
- **`-c`**, **`--charset`**: File containing the characters to use (encoded as UTF-8). Can also be
the name of one of the built-in charsets: ascii, ascii-extended, latin-0, latin-9, windows-1252,
extended. Default: ascii
- **`--compression-level`**: Compression level for generated PNG, from 0 to 9. Default: 9.
- **`--fast-pack`**: Enable fast packing algorithm for texture packager. Default: false.
- **`-h`**, **`--help`**: Show the help message.
- **`-v`**, **`--version`**: Show version.

#### Examples
```
java -jar msdfgen.jar -o output -t msdf -a sdf -s 42 -r 6 -c latin-9 font.ttf
```
Generate MSDF font files with SDF encoded in the alpha layer for input file `font.ttf`.
Font size is 42px and distance range is 6px. Charset is built-in *latin-9*.
The following image is created in the `output/` directory:

<img src="examples/roboto-msdfa.png" alt="MSDFA texture atlas"/><br>

---

```
java -jar msdfgen.jar -t msdf -a none -s 24 -r 4 -c charset.txt font1.ttf font2.ttf
```
Generate MSDF font files with no alpha layer for input files `font1.ttf` and `font2.ttf`.
Font size is 24px and distance range is 4px. Charset is taken from the `charset.txt` file.
The following images are created in the `output/` directory:

<img src="examples/roboto-bold-msdf.png" alt="MSDF texture atlas"/>  <img src="examples/satisfy-msdf.png" alt="MSDF texture atlas"/>

## Changelog
See [changelog](CHANGELOG.md).

## License
All code is licensed under [Apache License, Version 2.0](LICENSE).

## Contributing
Don't hesitate to open a pull request if you feel like something needs an improvement.
Just be sure to follow existing code style.

[msdfgen]: https://github.com/Chlumsky/msdfgen/releases
[releases]: https://github.com/maltaisn/msdf-gdx-gen/releases


CHANGELOG.md
### v0.1.4
- Generate whitespace characters (#6).

### v0.1.3
- Fixed out of memory error when generating large charset (#3)
- Added `--fast-pack` option to enable the fast mode for the texture packer.
- Added `mtsdf` field type matching the corresponding msdfgen field type.

### v0.1.2
- Fixed concurrent modification exception during generation (#2).

### v0.1.1
- Fixed `--texture-size` parameter not working (#1).
- Added `--version` parameter.
- Parameter values now printed before execution.

## v0.1.0
- **Initial release**


gen/src/main/kotlin/com/maltaisn/msdfgdx/gen/main.kt
/*
 * Copyright 2019 Nicolas Maltais
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.maltaisn.msdfgdx.gen

import com.beust.jcommander.JCommander
import com.maltaisn.msdfgdx.gen.BMFont.GenerationStep
import java.io.File
import java.text.DecimalFormat
import kotlin.math.floor
import kotlin.system.exitProcess


fun main(args: Array<String>) {
    val params = Parameters()
    val commander = JCommander.newBuilder().addObject(params).build()
    commander.programName = "msdfgen-bmfont"

    try {
        // Parse arguments
        try {
            commander.parse(*args)
        } catch (e: com.beust.jcommander.ParameterException) {
            paramError(e.message)
        }

        if (params.version) {
            // Get version from resources and print it.
            val versionRes = Parameters::class.java.classLoader.getResourceAsStream("version.txt")!!
            val version = String(versionRes.readBytes())
            println("msdf-gdx-gen v$version")
            exitProcess(0)
        }

        if (params.help) {
            // Show help message
            commander.usage()
            exitProcess(0)
        }

        // Validate arguments
        params.validate()

        // Show arguments
        println(params.summarize())
        println()

        // Generate bitmap fonts
        val numberFmt = DecimalFormat.getInstance()
        numberFmt.minimumFractionDigits = 1
        numberFmt.maximumFractionDigits = 1

        for (fontPath in params.params) {
            val startTime = System.currentTimeMillis()
            var stepStartTime = 0L

            val fontFile = File(fontPath)
            val bmfont = BMFont(fontFile, params)

            println("Generating distance field font for '${fontFile.name}'.")

            var lastPercentProgress = -1
            var lastUpdateTime = 0L
            bmfont.generate { step, progress ->
                val percentProgress = floor(progress * 100).toInt()
                if (percentProgress != lastPercentProgress || (System.currentTimeMillis() - lastUpdateTime) > 100) {
                    if (percentProgress == 0) stepStartTime = System.currentTimeMillis()
                    val stepDurationStr = numberFmt.format((System.currentTimeMillis() - stepStartTime) / 1000.0)

                    val stepName = when (step) {
                        GenerationStep.GLYPH -> "Generating glyph images"
                        GenerationStep.PACK -> "Packing glyphs into atlas"
                        GenerationStep.FONT_FILE -> "Generating BMFont file"
                        GenerationStep.COMPRESS -> "Compressing images"
                    }.padEnd(30, ' ')

                    // Print step name, show progress bar and percent progress.
                    print("\r$stepName[${"#".repeat((progress * 40).toInt())}" +
                            "${"-".repeat((40 - progress * 40).toInt())}] " +
                            "  $percentProgress % ($stepDurationStr s)")

                    if (percentProgress == 100) println()  // New line for next step
                    lastPercentProgress = percentProgress
                    lastUpdateTime = System.currentTimeMillis()
                }
            }

            // Done generating font, show duration.
            val durationStr = numberFmt.format((System.currentTimeMillis() - startTime) / 1000.0)
            println("DONE in $durationStr s\n")
        }

        exitProcess(0)

    } catch (e: ParameterException) {
        println("ERROR: ${e.message}\n")
        exitProcess(1)
    }
}


gen/src/main/kotlin/com/maltaisn/msdfgdx/gen/BMFont.kt
/*
 * Copyright 2019 Nicolas Maltais
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.maltaisn.msdfgdx.gen

import com.badlogic.gdx.Files
import com.badlogic.gdx.backends.lwjgl.LwjglFileHandle
import com.badlogic.gdx.graphics.Pixmap
import com.badlogic.gdx.graphics.g2d.TextureAtlas
import com.badlogic.gdx.tools.hiero.Kerning
import com.badlogic.gdx.tools.texturepacker.TexturePacker
import com.badlogic.gdx.utils.IntIntMap
import com.googlecode.pngtastic.core.PngImage
import com.googlecode.pngtastic.core.PngOptimizer
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.runBlocking
import java.awt.Canvas
import java.awt.Font
import java.awt.FontFormatException
import java.awt.font.FontRenderContext
import java.awt.geom.AffineTransform
import java.awt.geom.GeneralPath
import java.io.BufferedInputStream
import java.io.File
import java.io.FileInputStream
import java.util.concurrent.atomic.AtomicInteger
import kotlin.math.ceil
import kotlin.math.roundToInt

/**
 * Generate bitmap font files from a TTF [fontFile] with some [params].
 * Generated files are the texture atlas pages and a `.fnt` file (AngelCode BMFont specs, libGDX format).
 *
 * Note that font and glyph metrics seems kinda off sometimes. I used:
 * - https://github.com/libgdx/libgdx/tree/aff6dd4a2622d64a62196111eb018d4699a19c8a/extensions/gdx-tools/src/com/badlogic/gdx/tools/hiero
 * - https://github.com/soimy/msdf-bmfont-xml/blob/ff3669c2bfffd06f29bacedcdf3f073379b45e7e/index.js
 * To guess which font metrics would work, but it might not work for all fonts.
 */
class BMFont(
    private val fontFile: File,
    private val params: Parameters
) {

    private val font = try {
        Font.createFont(Font.TRUETYPE_FONT, fontFile).deriveFont(params.fontSize.toFloat())
    } catch (e: FontFormatException) {
        paramError("ERROR: Could not load font file: ${e.message}")
    }

    private val glyphs = sortedMapOf<Char, FontGlyph>()

    private val fontMetrics = Canvas().getFontMetrics(font)
    private val fontRenderContext = FontRenderContext(AffineTransform(), true, true)

    private var atlasData: TextureAtlas.TextureAtlasData? = null

    fun generate(progressListener: ProgressListener) {
        generateGlyphs(progressListener)
        pack(progressListener)
        generateFontFile(progressListener)
        compress(progressListener)
    }

    private fun generateGlyphs(progressListener: ProgressListener) {
        progressListener(GenerationStep.GLYPH, 0f)
        val glyphsGenerated = AtomicInteger()

        val kernings = Kerning().apply {
            load(fontFile.inputStream(), params.fontSize)
        }.kernings

        glyphs += runBlocking {
            // Create a new async job for each glyph to generate and await them all.
            params.charList.map { char ->
                GlobalScope.async {
                    val glyph = generateGlyph(char, kernings)

                    // Set progress
                    progressListener(GenerationStep.GLYPH,
                        glyphsGenerated.incrementAndGet().toFloat() / params.charList.length)

                    char to glyph
                }
            }.awaitAll()
        }

        progressListener(GenerationStep.GLYPH, 1f)
    }

    private fun generateGlyph(char: Char, kernings: IntIntMap): FontGlyph {
        // Create glyph vector and get its bounding box.
        val glyphVector = font.createGlyphVector(fontRenderContext, char.toString())
        val bounds = glyphVector.visualBounds

        // Character is printable
        val glyph = FontGlyph()

        // Set kerning distances
        for (other in params.charList) {
            val pair = (char.code shl 16) or other.code
            if (kernings[pair, 0] != 0) {
                glyph.kernings[other] = kernings[pair, 0]
            }
        }

        val pad = params.distanceRange / 2f
        val w = ceil(bounds.width + pad * 2).toInt()
        val h = ceil(bounds.height + pad * 2).toInt()

        // Get glyph path and translate it to center.
        val tx = -bounds.x + pad
        val ty = -bounds.y - bounds.height - pad
        val path = glyphVector.getGlyphOutline(0, tx.toFloat(), ty.toFloat()) as GeneralPath

        glyph.xOffset = (bounds.x - pad).roundToInt()
        glyph.yOffset = (fontMetrics.ascent + bounds.y - pad).roundToInt()
        glyph.xAdvance = glyphVector.getGlyphMetrics(0).advanceX.roundToInt()

        // Generate main glyph image.
        val gen = MsdfGen(params.msdfgen, w, h, params.distanceRange, Shape.fromPath(path).toString())
        val glyphImage = gen.generateImage(params.fieldType)

        glyph.width = w
        glyph.height = h
        glyph.image = glyphImage
        glyph.channels = if (params.hasAlphaChannel) FontGlyph.CHANNELS_RGBA else FontGlyph.CHANNELS_RGB

        if (params.alphaFieldType != Parameters.FIELD_TYPE_NONE && params.fieldType != Parameters.FIELD_TYPE_MTSDF) {
            // Generate glyph image used for alpha layer.
            // Then keep RGB channel of glyph image and use red channel of alpha image as alpha channel.
            val alphaImage = gen.generateImage(params.alphaFieldType)
            val glyphPixels = glyphImage.getRGB(0, 0, w, h, null, 0, w)
            val alphaPixels = alphaImage.getRGB(0, 0, w, h, null, 0, w)
            for (j in glyphPixels.indices) {
                glyphPixels[j] = (glyphPixels[j] and 0x00FFFFFF) or (alphaPixels[j] and 0xFF shl 24)
            }
            glyphImage.setRGB(0, 0, w, h, glyphPixels, 0, w)
        }

        return glyph

    }

    /**
     * Pack the generated glyphs to a texture atlas.
     */
    private fun pack(progressListener: ProgressListener) {
        val packer = TexturePacker(File(params.outputDir), TexturePacker.Settings().apply {
            paddingX = params.padding
            paddingY = params.padding

            maxWidth = params.textureSize[0]
            maxHeight = params.textureSize[1]
            alias = false
            ignoreBlankImages = false
            silent = true

            fast = params.fastPacking

            format = if (params.hasAlphaChannel) Pixmap.Format.RGBA8888 else Pixmap.Format.RGB888
        })
        val atlasFile = File(params.outputDir, fontFile.nameWithoutExtension + ".atlas")

        progressListener(GenerationStep.PACK, 0f)
        packer.setProgressListener(object : TexturePacker.ProgressListener() {
            override fun progress(progress: Float) {
                progressListener(GenerationStep.PACK, progress)
            }
        })

        // Add glyph images to packer
        for ((char, glyph) in glyphs) {
            packer.addImage(glyph.image, char.code.toString())
        }

        // Remove any existing atlas related files.
        // This is necessary because otherwise the packer tries to add new textures to the existing atlas.
        File(params.outputDir, fontFile.nameWithoutExtension + ".atlas").delete()
        var pageIndex = 0
        while (true) {
            val pageFile = getTextureAtlasFile(pageIndex)
            if (!pageFile.exists()) break
            pageFile.delete()
            pageIndex++
        }

        // Pack to atlas
        try {
            packer.pack(File(params.outputDir), fontFile.nameWithoutExtension)
        } catch (e: Exception) {
            println()
            paramError("Texture packer error: ${e.message}")
        }

        // Set the glyph positions and page from the generated atlas file
        val atlasData = TextureAtlas.TextureAtlasData(LwjglFileHandle(atlasFile, Files.FileType.Absolute),
            LwjglFileHandle("", Files.FileType.Absolute), false)
        for (region in atlasData.regions) {
            val char = region.name.toInt().toChar()
            val glyph = glyphs[char] ?: continue
            glyph.page = atlasData.pages.indexOf(region.page, true)
            glyph.x = region.left
            glyph.y = region.top
        }
        this.atlasData = atlasData

        // Delete atlas file, not used anymore.
        atlasFile.delete()
    }

    /**
     * Generate the font file from the generated `.atlas` file in the previous step.
     */
    private fun generateFontFile(progressListener: ProgressListener) {
        progressListener(GenerationStep.FONT_FILE, 0f)

        val bmfont = StringBuilder()
        val atlasData = checkNotNull(atlasData)

        // Info tag
        bmfont.appendLine("info face=\"${font.fontName}\" size=${params.fontSize} " +
                "bold=${if (font.isBold) 1 else 0} italic=${if (font.isItalic) 1 else 0} " +
                "charset=\"\" unicode=1 stretchH=100 smooth=1 aa=1 padding=0,0,0,0 spacing=0,0 outline=0")

        // Common tag
        bmfont.appendLine("common lineHeight=${fontMetrics.height} " +
                "base=${fontMetrics.ascent} scaleW=${params.textureSize[0]} " +
                "scaleH=${params.textureSize[1]} pages=${atlasData.pages.size} " +
                "packed=0 alphaChnl=0 redChnl=0 greenChnl=0 blueChnl=0 " +
                "distanceRange=${params.distanceRange}")

        // Page tags
        for (i in 0 until atlasData.pages.size) {
            bmfont.appendLine("page id=$i file=\"${getTextureAtlasFile(i).name}\"")
        }

        val kerningsCount = glyphs.values.sumOf { it.kernings.size }
        val elementsCount = glyphs.size + kerningsCount
        var elementsDone = 0f

        // Char tags
        bmfont.appendLine("chars count=${glyphs.size}")
        for ((char, glyph) in glyphs) {
            bmfont.appendLine("char id=${char.code} x=${glyph.x} y=${glyph.y} " +
                    "width=${glyph.width} height=${glyph.height} " +
                    "xoffset=${glyph.xOffset} yoffset=${glyph.yOffset} " +
                    "xadvance=${glyph.xAdvance} page=${glyph.page} chnl=${glyph.channels}")
            elementsDone++
            progressListener(GenerationStep.FONT_FILE, elementsDone / elementsCount)
        }

        // Kerning tags
        bmfont.appendLine("kernings count=$kerningsCount")
        for ((char, glyph) in glyphs) {
            for ((other, kerning) in glyph.kernings) {
                bmfont.appendLine("kerning first=${char.code} second=${other.code} amount=$kerning")
                elementsDone++
                progressListener(GenerationStep.FONT_FILE, elementsDone / elementsCount)
            }
        }

        // Write to file
        File(params.outputDir, fontFile.nameWithoutExtension + ".fnt")
            .writeText(bmfont.toString())
    }

    private fun compress(progressListener: ProgressListener) {
        if (params.compressionLevel == 0) {
            // No compression
            return
        }
        progressListener(GenerationStep.COMPRESS, 0f)

        val atlasData = checkNotNull(atlasData)
        val pngOptimizer = PngOptimizer()
        for (i in 0 until atlasData.pages.size) {
            val file = getTextureAtlasFile(i).absolutePath
            val inputStream = BufferedInputStream(FileInputStream(file))
            inputStream.use {
                val pngImage = PngImage(inputStream)
                pngImage.fileName = file
                pngOptimizer.optimize(pngImage, file, false, params.compressionLevel)
            }
            progressListener(GenerationStep.COMPRESS, (i + 1).toFloat() / atlasData.pages.size)
        }
    }

    private fun getTextureAtlasFile(pageIndex: Int) = File(params.outputDir, fontFile.nameWithoutExtension +
            (if (pageIndex == 0) "" else (pageIndex + 1).toString()) + ".png")

    enum class GenerationStep {
        GLYPH,
        PACK,
        FONT_FILE,
        COMPRESS
    }

}

typealias ProgressListener = (step: BMFont.GenerationStep, progress: Float) -> Unit


gen/src/main/kotlin/com/maltaisn/msdfgdx/gen/MsdfGen.kt
/*
 * Copyright 2019 Nicolas Maltais
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.maltaisn.msdfgdx.gen

import java.awt.image.BufferedImage
import java.io.BufferedReader
import java.io.InputStreamReader


class MsdfGen(private val msdfgen: String,
              private var width: Int,
              private var height: Int,
              private var distanceRange: Int,
              private var shapeDescr: String) {

    /**
     * Generate the glyph image for the specified shape, dimensions and distance range.
     * The output is an image with [BufferedImage.TYPE_INT_ARGB] format.
     */
    fun generateImage(fieldType: String): BufferedImage {
        val process = ProcessBuilder().command(
                msdfgen,
                "-format", "text",
                "-stdout",
                "-size", width.toString(), height.toString(),
                "-pxrange", distanceRange.toString(),
            "-defineshape", shapeDescr,
            fieldType)
            .start()

        // Convert text output to byte array.
        val reader = BufferedReader(InputStreamReader(process.inputStream))
        val pixels = IntArray(width * height)

        var line: String? = reader.readLine()
        val channels = (line!!.length + 1) / width / 3
        check(channels == 1 || channels == 3 || channels == 4) {
            "msdfgen generated image with unsupported channels count."
        }

        // Output is lines of hex-encoded bytes separated by spaces, one for each image line.
        // There are either 1, 3, or 4 bytes per pixel. Y coordinate must be reversed.
        var i = pixels.size
        while (line != null) {
            i -= width
            for (j in 0 until width) {
                // Build pixel width 3 consecutive bytes.
                val pos = j * channels * 3
                val a: Int
                val r: Int
                val g: Int
                val b: Int
                when (channels) {
                    1 -> {
                        r = line.substring(pos, pos + 2).toInt(16)
                        g = r
                        b = r
                        a = 255
                    }
                    3 -> {
                        r = line.substring(pos, pos + 2).toInt(16)
                        g = line.substring(pos + 3, pos + 5).toInt(16)
                        b = line.substring(pos + 6, pos + 8).toInt(16)
                        a = 255
                    }
                    4 -> {
                        r = line.substring(pos, pos + 2).toInt(16)
                        g = line.substring(pos + 3, pos + 5).toInt(16)
                        b = line.substring(pos + 6, pos + 8).toInt(16)
                        a = line.substring(pos + 9, pos + 11).toInt(16)
                    }
                    else -> error("")
                }
                pixels[i + j] = (a shl 24) or (r shl 16) or (g shl 8) or b
            }
            line = reader.readLine()
        }

        // Build image from array of pixels.
        val image = BufferedImage(width, height, BufferedImage.TYPE_INT_ARGB)
        image.setRGB(0, 0, width, height, pixels, 0, width)
        return image
    }

}


gen/src/main/kotlin/com/maltaisn/msdfgdx/gen/Parameters.kt
/*
 * Copyright 2019 Nicolas Maltais
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.maltaisn.msdfgdx.gen

import com.beust.jcommander.Parameter
import com.beust.jcommander.Parameters
import java.io.File
import java.io.IOException


@Parameters(separators = " =")
class Parameters {

    @Parameter
    var params: List<String> = mutableListOf()

    @Parameter(names = ["-g", "--msdfgen"], description = "Path of the msdfgen executable", order = 0)
    var msdfgen: String = "msdfgen.exe"

    @Parameter(names = ["-o", "--output"], description = "Output path of generated font textures", order = 1)
    var output: String? = null

    @Parameter(names = ["-t", "--field-type"], description = "Field type: sdf | psdf | msdf | mtsdf", order = 2)
    var fieldType: String = FIELD_TYPE_MSDF

    @Parameter(names = ["-a", "--alpha-field-type"], description = "Alpha field type: none | sdf | psdf", order = 3)
    var alphaFieldType: String = ""

    @Parameter(names = ["-s", "--font-size"], description = "Font size for generated textures", order = 4)
    var fontSize: Int = 32

    @Parameter(names = ["-r", "--distance-range"], description = "Distance range in which SDF is encoded", order = 5)
    var distanceRange: Int = 5

    @Parameter(names = ["-d", "--texture-size"], arity = 2, description = "Maximum width and height of generated atlas pages", order = 6)
    var textureSize: List<Int> = listOf(512, 512)

    @Parameter(names = ["-p", "--padding"],
        description = "Padding between glyphs and on the border of the atlas pages",
        order = 7)
    var padding: Int = 2

    @Parameter(names = ["-c", "--charset"], description = "File containing the characters to use (encoded as UTF-8). " +
            "Can also be one of: ascii, ascii-extended, latin-0, latin-9, windows-1252, extended.", order = 8)
    var charset: String = "ascii"

    @Parameter(names = ["--compression-level"],
        description = "Compression level for generated PNG, from 0 to 9",
        order = 9)
    var compressionLevel = 9

    @Parameter(names = ["--fast-pack"],
        description = "Whether to use the faster but less efficient packing algorithm",
        order = 10)
    var fastPacking: Boolean = false

    @Parameter(names = ["-h", "--help"], description = "Show help message", help = true, order = 11)
    var help = false

    @Parameter(names = ["-v", "--version"], description = "Show version", order = 12)
    var version = false

    /** List of characters from charset. */
    var charList = ""
        private set

    /** Output directory. */
    var outputDir = System.getProperty("user.dir")
        private set

    /**
     * Whether output bitmap font has glyphs encoded in alpha channel or not.
     */
    val hasAlphaChannel: Boolean
        get() = alphaFieldType != FIELD_TYPE_NONE || fieldType == FIELD_TYPE_MTSDF

    /**
     * Validate arguments
     */
    fun validate() {
        // Validate input files
        if (params.isEmpty()) paramError("No input file.")
        for (inputFile in params) {
            if (!File(inputFile).exists()) {
                paramError("Input file '$inputFile' doesn't exist.")
            }
        }

        // Create output directory
        if (output != null) outputDir = output
        val outputPath = File(outputDir)
        outputPath.mkdirs()
        if (!outputPath.exists()) paramError("Could not create output directory at '$outputDir'.")

        // Check if msdfgen executable can be executed.
        try {
            Runtime.getRuntime().exec(msdfgen)
        } catch (e: IOException) {
            paramError("msdfgen executable '$msdfgen' doesn't exist or isn't executable.")
        }

        // Due to a JCommander bug, a list parameter default value is not overwritten, only appended.
        // https://github.com/cbeust/jcommander/issues/137
        if (textureSize.size == 4) {
            textureSize = textureSize.subList(2, 4)
        }

        // Validate other arguments
        when {
            fieldType !in listOf(FIELD_TYPE_SDF, FIELD_TYPE_PSDF, FIELD_TYPE_MSDF, FIELD_TYPE_MTSDF) ->
                paramError("Invalid field type '$fieldType'")
            alphaFieldType !in listOf(FIELD_TYPE_NONE, FIELD_TYPE_SDF, FIELD_TYPE_PSDF) ->
                paramError("Invalid field type '$alphaFieldType'")
            fontSize < 8 -> paramError("Font size must be at least 8.")
            distanceRange < 1 -> paramError("Distance range must be at least 1.")
            textureSize.any { d -> d !in VALID_TEXTURE_SIZES } -> paramError("Texture size must be power of two between 32 and 65536.")
            padding < 0 -> paramError("Padding must be at least 0.")
        }

        if (alphaFieldType != FIELD_TYPE_NONE && fieldType == FIELD_TYPE_MTSDF) {
            // if using mtsdf, msdfgen will take care of generating the alpha channel, so alpha field type has no effect.
            println("WARNING: alpha field type is ignored when using mtsdf field type.")
        }

        // Get charset from file or builtin
        // Also remove duplicate characters and sort them.
        charList = BUILTIN_CHARSETS[charset] ?: try {
            File(charset).readText()
        } catch (e: IOException) {
            paramError("Could not read charset file.")
        }.toSortedSet().joinToString("")
    }

    /**
     * Return a summary of the parameter values.
     */
    fun summarize() = """
        |Font files:
        |  - ${params.joinToString("\n  - ") { File(it).absolutePath }}
        |Output: ${File(outputDir).absolutePath}
        |Field type: $fieldType
        |Alpha field type: $alphaFieldType
        |Font size: $fontSize px
        |Distance range: $distanceRange px
        |Texture size: ${textureSize[0]} x ${textureSize[1]} px
        |Padding: $padding px
        |Charset: ${charList.length} chars
        |Compression level: $compressionLevel
        |Fast packing: $fastPacking
    """.trimMargin()

    companion object {
        const val FIELD_TYPE_NONE = "none"
        const val FIELD_TYPE_SDF = "sdf"
        const val FIELD_TYPE_PSDF = "psdf"
        const val FIELD_TYPE_MSDF = "msdf"
        const val FIELD_TYPE_MTSDF = "mtsdf"

        private val VALID_TEXTURE_SIZES = List(12) { 1 shl (it + 5) }

        private val BUILTIN_CHARSETS = mapOf(
            "test" to " A@jp&ÂO!-$",
            /* ASCII printable chars up to 127. */
            "ascii" to " !\"#\$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~",
            /* ASCII printable chars up to 255 minus box chars and some math. */
            "ascii-extended" to " !\"#\$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|} ~ÇüéâäàåçêëèïîìÄÅÉæÆôöòûùÿÖÜ¢£¥₧ƒáíóúñÑªº¿⌐¬½¼¡«»αßΓπΣσµτΦΘΩδ∞φε∩≡±≥≤÷≈°∙·√ⁿ²■",
            /* ISO/IEC 8859-1 aka latin-0. */
            "latin-0" to " !\"#\$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬\u00AD®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ",
            /* ISO/IEC 8859-15 aka latin-9. */
            "latin-9" to " !\"#\$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£€¥Š§š©ª«¬\u00AD®¯°±²³Žµ¶·ž¹º»ŒœŸ¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿ",
            /* windows-1252 (superset of latin-0). */
            "windows-1252" to " !\"#\$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬\u00AD®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿŒœŠšŸŽžƒˆ˜–—‘’‚“”„†‡•…‰‹›€™",
            /* Hiero's extended charset. */
            "extended" to " !\"#\$%&'()*+,-./0123456789:;<=>?@ABCDEFGHIJKLMNOPQRSTUVWXYZ[\\]^_`abcdefghijklmnopqrstuvwxyz{|}~ ¡¢£¤¥¦§¨©ª«¬\u00AD®¯°±²³´µ¶·¸¹º»¼½¾¿ÀÁÂÃÄÅÆÇÈÉÊËÌÍÎÏÐÑÒÓÔÕÖ×ØÙÚÛÜÝÞßàáâãäåæçèéêëìíîïðñòóôõö÷øùúûüýþÿĀāĂăĄąĆćĈĉĊċČčĎďĐđĒēĔĕĖėĘęĚěĜĝĞğĠġĢ" +
                    "ģĤĥĦħĨĩĪīĬĭĮįİıĲĳĴĵĶķĸĹĺĻļĽľĿŀŁłŃńŅņŇňŉŊŋŌōŎŏŐőŒœŔŕŖŗŘřŚśŜŝŞşŠšŢţŤťŦŧŨũŪūŬŭŮůŰűŲųŴŵŶŷŸŹźŻżŽžſʹ͵ͺͻͼͽ;΄΅Ά·ΈΉΊΌΎΏΐΑΒΓΔΕΖΗΘΙΚΛΜΝΞΟΠΡΣΤΥΦΧΨΩΪΫάέήίΰαβγδεζηθικλμνξοπρςστυφχψωϊϋόύώϐϑϒϓϔϕϖϗϘϙϚϛϜϝϞϟϠϡϢϣϤϥϦϧϨϩϪϫϬϭϮϯϰϱϲϳϴϵ϶ϷϸϹϺϻϼϽϾϿЀЁЂЃЄЅІЇЈЉЊЋЌЍЎЏАБВГДЕ" +
                    "ЖЗИЙКЛМНОПРСТУФХЦЧШЩЪЫЬЭЮЯабвгдежзийклмнопрстуфхцчшщъыьэюяѐёђѓєѕіїјљњћќѝўџѠѡѢѣѤѥѦѧѨѩѪѫѬѭѮѯѰѱѲѳѴѵѶѷѸѹѺѻѼѽѾѿҀҁ҂҃҄҅҆҇҈҉ҊҋҌҍҎҏҐґҒғҔҕҖҗҘҙҚқҜҝҞҟҠҡҢңҤҥҦҧҨҩҪҫҬҭҮүҰұҲҳҴҵҶҷҸҹҺһҼҽҾҿӀӁӂӃӄӅӆӇӈӉӊӋӌӍӎӏӐӑӒӓӔӕӖӗӘәӚӛӜӝӞӟӠӡӢӣӤӥӦӧӨөӪӫӬӭӮӯӰӱӲӳӴӵӶӷӸӹӺӻӼӽӾӿԀԁԂԃԄԅԆԇԈԉԊԋԌ" +
                    "ԍԎԏԐԑԒԓԔԕԖԗԘԙԚԛԜԝԞԟԠԡԢԣԤԥԦԧ           \u200B\u200C\u200D\u200E\u200F‒–—―‖‗‘’‚‛“”„‟†‡•…\u202A\u202B\u202C\u202D\u202E ‰′″‴‹›‼‾⁄⁞\u206A\u206B\u206C\u206D\u206E\u206F₠₡₢₣₤₥₦₧₨₩₫€₭₮₯₰₱₲₳₴₵₹₺ⱠⱡⱢⱣⱤⱥⱦⱧⱨⱩⱪⱫⱬⱭⱱⱲⱳⱴⱵⱶⱷ"
        )
    }

}
